index.js

```diff
@@ -277,7 +278,13 @@ RedirectableRequest.prototype._performRequest = function () {
   // Create the native request
   var request = this._currentRequest =
         nativeProtocol.request(this._options, this._onNativeResponse);
-  this._currentUrl = url.format(this._options);
+  if (path.isAbsolute(this._options.path)) {
```

Comment:
`path` is for file paths, so that is definitely not right.

Does this case only happen with proxies? Does it make sense to detect whether a proxy is active?

Reply:
The RFC is clear about the 3 forms for the `request URI` part of the request line:
1) `*`: in requests that do not pertain to a specific resource on the server (e.g. `OPTIONS * HTTP/1.1` ) - unlikely to be relevant for redirection scenarios.
2) An absolute URI: to access a resource on a remote server <-- indirect request through proxy
3) A **absolute path** (emphasis mine: this is how the RFC describes this case): to access a resource on the target server <-- "regular", direct request

`path` is for file paths, though `path` itself is "OS aware". An alternative would be to use `path.posix`, which to guarantee using an API with the same separators and same rules about what constitutes an absolute path, though my tests pass on Windows, in cmd. A worse option would be string parsing and string building.

Re: your question about proxies, I don't think you can detect proxy usage as such, I would just focus on following the HTTP spec: if the request line included an absolute path, a relative redirect needs to be built based on that absolute path, not the current host. Po-tah-tos, po-tay-tos, you might say, but I think that would make this part of the code clearer.

Reply:
If you do feel very strongly against using `path.isAbsolute`, I guess the only option is to essentially reimplement `path.isAbsolute` using regular expressions or string parsing, which sounds much riskier, or use an external library, which increases the potential for vulnerabilities and bugs and just seems overkill to me.

Reply:
> *: in requests that do not pertain to a specific resource on the server (e.g. OPTIONS * HTTP/1.1 ) - unlikely to be relevant for redirection scenarios.

Ooh yes, we'll need to cover that case too, or we'll be appending stars to URLs.

---

Specifically, the RFC says:

> When making a request directly to an origin server, other than a
   CONNECT or server-wide OPTIONS request (as detailed below), a client
   MUST send only the absolute path and query components of the target
   URI as the request-target.

and

> When making a request to a proxy, other than a CONNECT or server-wide
   OPTIONS request (as detailed below), a client MUST send the target
   URI in absolute-form as the request-target.

So your statement on absolute path detection is correct.

----

> path is for file paths, though path itself is "OS aware"

That's what's so dangerous about it; what we do with HTTP is platform-neutral.

> I guess the only option is to essentially reimplement path.isAbsolute

No, we implement an absolute URI path check, which is different. That check is `/^\//.test(uri)` as per

> `absolute-path = 1*( "/" segment )`

Reply:
PS I suggest to quote/link the relevant spec parts in code comments, as you'll see in other parts of the code.

Reply:
> Ooh yes, we'll need to cover that case too, or we'll be appending stars to URLs.

I don't think it'd be a problem, since passing `*` as the resource target means the origin server itself. I don't think this can ever return a redirection response, so what is stored in `this._currentUrl` is more or less important as it won't be used to build a subsequent URL. Besides, there's no way to represent this as a URL, really. 

In any case, I suggest not implementing anything special for `*`. Alternatively, you might want to replace `this._currentUrl` with `this._currentRequestTarget` that would be an object with unambiguous `host`, `port`, and `path` properties - basically, a subset of `http.ClientRequest`. That would let you build redirect URLs (or redirect `http.ClientRequest`s) unambiguously and with fewer risks of errors than relying on `url.resolve`, `url.parse` and `url.format`.

... that could all be a future PR :smile:

---

> No, we implement an absolute URI path check, which is different. That check is `/^\//.test(uri)` as per
> `absolute-path = 1*( "/" segment )`

RFC 2068, though outdated, actually defined this as 
> "/" rel_path

but in any case, both definitions are consistent and the check is indeed pretty safe/simple. So, sounds good! I'll work on that tonight and push an update. Can you clarify which RFC specifically you'd like me to quote?

Reply:
Good ol' RFC7230. Thanks a bunch!

Reply:
Comments added in https://github.com/follow-redirects/follow-redirects/pull/199/commits/143ee38a322ab111b83f84ea398ee78d8acf6bdd

Check replaced with regex in https://github.com/follow-redirects/follow-redirects/pull/199/commits/6451d6c7b0811f033a23bf1f8ba29fa745ce511a

Reply:
I left a comment abbout `OPTIONS` and `CONNECT`, for which I wouldn't expect servers to return 3xx responses. 

In any case, I still think a better and somewhat clearer option, long term, would be to create a copy of the subset of `this._options` for use in computing redirect options rather than `this._currentUrl`, which works fine with `origin-form` and `absolute-form`, but are non-sensical for `asterisk-form` and `authority-form`. I leave it up to you if this PR as-is is fine or if you would like me to make this change too.

Thanks for the quick feedback, by the way!